[PWGJE] First update heavy flavour tagging analysis

hanseopark opened the pull request:
Core/JetTaggingUtilities.h
- Add calculation of geometric sign which is product dot jet momentum and dca momentum
- 
DataModel/JetTagging.h
- Add propagation of track for dca_X and dcaY
- Add table of tracks which have parameter for track counting. it is expected to expend different method

TableProducer/jettaggerhf.cxx
- Making table for tagging jet and tracks

TableProducer/jettaggerhfExtension.cxx
- Add propagation of track for dca_X and dca_Y

Tasks/jettaggerhfTC.cxx
- First QA for hf jet tagging for flavour decision and comparing track's parameters

nzardosh commented on PWGJE/Core/JetTaggingUtilities.h:
```diff
@@ -38,6 +39,7 @@ enum JetTaggingSpecies {
 
 namespace JetTaggingUtilities
 {
+const int cmTomum = 10000;
```
can you put a comment here of what this is please?

hanseopark replied:
const int cmTomum = 10000; // using cm -> #mum for impact parameter (dca)

I add a comment

nzardosh commented on PWGJE/Core/JetTaggingUtilities.h:
```diff
@@ -275,6 +278,22 @@ int jetOrigin(T const& jet, U const& particles, float dRMax = 0.25)
+int getGeoSign(T const& collision, U const& jet, V const& track)
+{
+  auto sign = TMath::Sign(1, (track.dcaX() - collision.posX()) * jet.px() + (track.dcaY() - collision.posY()) * jet.py() + (track.dcaZ() - collision.posZ()) * jet.pz());
+  if (sign < -1 || sign > 1)
```
what is this if for?

hanseopark replied:
/**
 * return geometric sign which is calculated scalar product between jet axis with DCA (track propagated to PV )
    * positive and negative value are expected from primary vertex
    * positive value is expected from secondary vertex
    *
    * @param collision which is needed external table of collision due to postion X and Y
    * @param jet
    * @param track which is needed each DCA_X and Y which is measured in jettaggerhfExtension.cxx
    */

nzardosh commented on PWGJE/DataModel/JetTagging.h:
```diff
@@ -26,28 +27,143 @@ using namespace o2::analysis;
+                    _name_##tagging::Algorithm2,                 \
+                    _name_##tagging::Algorithm3);
+
+#define DECLARE_TAGTRACK_TABLE(_jet_type_, _name_, _track_type_, _description_)  \
```
please use TRACKTAG_TABLE_DEF as the name for consistency

nzardosh replied:
same as instances below of other defines

nzardosh commented on PWGJE/TableProducer/jettaggerhf.cxx:
```diff
@@ -21,64 +22,113 @@
 using namespace o2::framework::expressions;
 
-template <typename JetTableData, typename JetTableMCD, typename JetTaggingTableData, typename JetTaggingTableMCD>
+template <typename JetTableData, typename JetTableMCD, typename JetTaggingTableData, typename ConstTaggingTableData, typename ConstTCTaggingTableData, typename JetTaggingTableMCD, typename ConstTaggingTableMCD, typename ConstTCTaggingTableMCD>
```
why do the new ones have const in the name?

nzardosh commented on PWGJE/TableProducer/jettaggerhf.cxx:
```diff
@@ -21,64 +22,113 @@
   PROCESS_SWITCH(JetTaggerHFTask, processDummy, "Dummy process", true);
 
-  void processData(aod::JCollision const& collision, JetTableData const& jets, aod::JTracks const& tracks)
+  void processData(soa::Join<aod::JCollisions, aod::JCollisionPIs>::iterator const& jcollision, aod::Collisions&, JetTableData const& jets, aod::JTracks&, JetTagTracksData const& tracks)
```
please here use the predefined "using: keyowrds found at the bottom of DataModel/Jet.h

nzardosh commented on PWGJE/TableProducer/jettaggerhf.cxx:
```diff
@@ -21,64 +22,113 @@
+        vecSignedIP3Ds.push_back(geoSign * TMath::Abs(track.dcaXYZ()) / TMath::Sqrt(track.sigmaDcaXYZ2()) * JetTaggingUtilities::cmTomum);
+      }
+      constTableData(taggingTableData.lastIndex(), trackconst);
+      constTCTableData(taggingTableData.lastIndex(), 0, jet.pt(), jet.eta(), jet.phi(), vecGeoSign, vecTrackPt, vecTrackEta, vecTrackPhi, vecSignedIP2D, vecSignedIP2Ds, vecSignedIP3D, vecSignedIP3Ds);
```
why do you save the jet information for each track here? I think its better to remove this since it looks like this table is anyways joinable to the jet table right?

nzardosh commented on PWGJE/TableProducer/jettaggerhf.cxx:
```diff
@@ -21,64 +22,113 @@
+      for (auto& track : jet.template tracks_as<JetTagTracksData>()) {
+        trackconst.push_back(track.globalIndex());
+        int geoSign = JetTaggingUtilities::getGeoSign(collision, jet, track);
+        trackconst.push_back(track.globalIndex());
```
why is the track globalIndex pushed back twice?

nzardosh commented on PWGJE/TableProducer/jettaggerhf.cxx:
```diff
@@ -21,64 +22,113 @@
+        trackconst.push_back(track.globalIndex());
+        vecGeoSign.push_back(geoSign);
+        vecTrackPt.push_back(track.pt());
+        vecTrackEta.push_back(track.eta());
```
this is fine if you intend to store these tables later for offline processing. If not you can just put an index to the track table as you have above aand then you dont need to save pt, eta and phi of tracks

nzardosh commented on PWGJE/DataModel/JetTagging.h:
```diff
@@ -26,28 +27,143 @@ using namespace o2::analysis;
+DECLARE_SOA_COLUMN(JetPt, jetPt, float);
+DECLARE_SOA_COLUMN(JetEta, jetEta, float);
+DECLARE_SOA_COLUMN(JetPhi, jetPhi, float);
+DECLARE_SOA_COLUMN(VecGeoSign, vecGeoSign, std::vector<int>);
```
you can check if there is already a predefined array coloumn type or not

nzardosh replied:
No not an index coloumn, but just an array one. But possible there isnt

nzardosh commented on PWGJE/TableProducer/jettaggerhf.cxx:
```diff
@@ -21,64 +22,113 @@
+        vecSignedIP3D.push_back(track.dcaXYZ() / TMath::Sqrt(track.sigmaDcaXYZ2()));
+        vecSignedIP3Ds.push_back(geoSign * TMath::Abs(track.dcaXYZ()) / TMath::Sqrt(track.sigmaDcaXYZ2()));
+      }
+      constTableMCD(taggingTableMCD.lastIndex(), trackconst);
```
since these tables can be joined to the taggingTableMCD I dont think you need to the indices. you can just  join them later

nzardosh commented on PWGJE/TableProducer/jettaggerhfExtension.cxx:
```diff
@@ -0,0 +1,197 @@
+    runNumber = bc.runNumber();
+  }
+
+  void calculateDcaXYZ(float& dcaXYZ, float& sigmaDcaXYZ2, float dcaXY, float dcaZ, float cYY, float cZY, float cZZ, float sigmaDcaXY2, float sigmaDcaZ2)
```
does this not exist already in recodecay?

nzardosh commented on PWGJE/TableProducer/jettaggerhfExtension.cxx:
```diff
@@ -0,0 +1,197 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+// Task to produce a table joinable to the jet tables for hf jet tagging
```
please fix this

nzardosh commented on PWGJE/TableProducer/jettaggerhfExtension.cxx:
```diff
@@ -0,0 +1,197 @@
+    o2::base::Propagator::initFieldFromGRP(grpmag);
+    o2::base::Propagator::Instance()->setMatLUT(lut);
+    mVtx = ccdb->getForTimeStamp<o2::dataformats::MeanVertexObject>(mVtxPath, bc.timestamp());
+    runNumber = bc.runNumber();
```
why is this needed?

nzardosh commented on PWGJE/TableProducer/jettaggerhfExtension.cxx:
```diff
@@ -0,0 +1,197 @@
+    LOG(info) << "Setting magnetic field to current " << grpmag->getL3Current() << " A for run " << bc.runNumber() << " from its GRPMagField CCDB object";
+    o2::base::Propagator::initFieldFromGRP(grpmag);
+    o2::base::Propagator::Instance()->setMatLUT(lut);
+    mVtx = ccdb->getForTimeStamp<o2::dataformats::MeanVertexObject>(mVtxPath, bc.timestamp());
```
it seems in this function you only need this line?

hanseopark replied:
I just copy from Common/TableProducer/trackPropagation.cxx
it doesn't need to set magnetic field from ccdb?

nzardosh commented on PWGJE/TableProducer/jettaggerhfExtension.cxx:
```diff
@@ -0,0 +1,197 @@
+      auto xyz = trackPar.getXYZGlo();
+      float dcaX = 999.;
+      float dcaY = 999.;
+      if (track.has_collision()) {
```
isnt this already checked in line 131?

hanseopark replied:
is it right. I will change

nzardosh commented on PWGJE/TableProducer/jettaggerhfExtension.cxx:
```diff
@@ -0,0 +1,197 @@
+  }
+  PROCESS_SWITCH(JetTaggerHFTrackExtension, processDummy, "Dummy process", true);
+
+  void processTracks(soa::Join<aod::Tracks, aod::TracksCov, aod::TrackSelection, aod::TracksDCA, aod::TracksDCACov, aod::McTrackLabels>::iterator const& track, aod::BCsWithTimestamps&, aod::Collisions&)
```
what exactly is this task doing? It seems to be recalculating the DCA of tracks but it is doing so with respect to the original collision? dont these coloumns already exist?

nzardosh reviewed (commented):
=

hanseopark reviewed (commented):
=

hanseopark commented:
I have changed many things which are remove constituents in jet tags, propagation part, add comment and use pre-defined table definition.

DataModel/JetTagging.h
I add extendion of DCA for JTracks becaues I want to use extend from JTracks directly not using originaltrack. I'm not sure why we shouldn't when we can extend JTracks by simply calling the track. Is it much of a memory disadvantage?
If this is possible, I would also like to extend JCollision to use posX and posY as well (stored in the tag data model).

TableProducer/jettaggerhf.cxx
There are saved with flavour information of jet and table for ML.

TableProducer/jettaggerhfExtension.cxx
I have changed that I will not use propagation seperatively because when we use getXYZGlo(), it can be called for propagated track.
So, I did extend JTracks in here.

Task/jettaggerhfTC.cxx
There are task of general QA for hf tagger and analysis of track counting method for prototype.

Let me know if you have any comments

nzardosh commented on PWGJE/TableProducer/jettaggerhf.cxx:
```diff
@@ -12,73 +12,129 @@
 struct JetTaggerHFTask {
 
   Produces<JetTaggingTableData> taggingTableData;
+  Produces<StoredJetTaggingTableData> storedTaggingTableData;
```
I think you can change the name of this to something else. Since both this and the one above it are stored. Also the logic seems a bit strange. You have the algorithm outputs in jettaggingtabledata and then you have some variable outputs in storedjettaggingtabledata. However it seems these should be two different tasks and that you should be able to run the trained ml model in this task
